Issue #498: Implement a simple print function for forecast objects.

toshiakiasakura opened the pull request:
## Description

This PR closes #498, and a refresh version of the PR from #584. 

The updated simple print function behaviour becomes like the following. 
`get_protected_columns` function returns columns including "scores columns" and then I excluded score columns from `protected columns` field to avoid redunduncy. 

```r
example_quantile %>%
  set_forecast_unit(c("location", "target_end_date", "target_type", "horizon", "model")) %>%
  as_forecast() %>%
  print
#>Forecast type:
#>[1] "quantile"
#>
#>Protected columns:
#>[1] "observed"  "quantile"  "predicted"
#>
#>Forecast units:
#>[1] "location"        "target_end_date" "target_type"     "horizon"        
#>[5] "model"          
#>
#>       observed quantile predicted location target_end_date target_type horizon
#>    1:   127300       NA        NA       DE      2021-01-02       Cases      NA
#>    2:     4534       NA        NA       DE      2021-01-02      Deaths      NA
#>    3:   154922       NA        NA       DE      2021-01-09       Cases      NA
#>    4:     6117       NA        NA       DE      2021-01-09      Deaths      NA
#>    5:   110183       NA        NA       DE      2021-01-16       Cases      NA
#>   ---                                                                         
#>20541:       78    0.850       352       IT      2021-07-24      Deaths       2
```
and 

```r
example_quantile %>%
  set_forecast_unit(c("location", "target_end_date", "target_type", "horizon", "model")) %>%
  as_forecast() %>%
  add_coverage() %>%
  print
#>Forecast type:
#>[1] "quantile"
#>
#>Protected columns:
#>[1] "observed"  "quantile"  "predicted" "range"    
#>
#>Score columns (Protected):
#>[1] "interval_coverage"           "interval_coverage_deviation"
#>[3] "quantile_coverage"           "quantile_coverage_deviation"
#>
#>Forecast units:
#>[1] "location"        "target_end_date" "target_type"     "horizon"        
#>[5] "model"          
#>
#>       observed quantile predicted location target_end_date target_type horizon
#>    1:   127300       NA        NA       DE      2021-01-02       Cases      NA
#>    2:     4534       NA        NA       DE      2021-01-02      Deaths      NA
#>    3:   154922       NA        NA       DE      2021-01-09       Cases      NA
#>    4:     6117       NA        NA       DE      2021-01-09      Deaths      NA
#>    5:   110183       NA        NA       DE      2021-01-16       Cases      NA
#>   ---                                                                         
#>20541:       78    0.950       370       IT      2021-07-24      Deaths       3
```


## Checklist

- [x] My PR is based on a package issue and I have explicitly linked it.
- [x] I have included the target issue or issues in the PR title as follows: *issue-number*: PR title
- [x] I have tested my changes locally.
- [ ] I have added or updated unit tests where necessary.
- [ ] I have updated the documentation if required.
- [x] I have built the package locally and run rebuilt docs using roxygen2.
- [x] My code follows the established coding standards and I have run `lintr::lint_package()` to check for style issues introduced by my changes. 
- [ ] I have added a news item linked to this PR.
- [ ] I have reviewed CI checks for this PR and addressed them as far as I am able.

<!-- Thanks again for this PR - @scoringutils dev team -->

nikosbosse commented:
Nice! Thanks so much and apologies for taking so long to get back to you. 
Looking good overall, some thoughts and comments: 

I think the following line
https://github.com/toshiakiasakura/scoringutils/blob/d4d0446a09ace4ec5bc509374ac5e79ca85367fb/R/available_forecasts.R#L63
needs to be changed to 
``` 
out <- as.data.table(data)[, .(count = .N), by = by]
```
The reason is that the object is still of class `forecast_*()`, but `get_forecast_counts()` removes the columns `observed` and `predicted` which can lead to an error. 

To make sure we have that covered in the future we should in addition add a test to `test-available_forecasts.R`, something like 
```
test_that("get_forecast_counts() works on an object of class `forecast_*`", {
  ex <- suppressMessages(as_forecast(example_integer))
  expect_no_condition(
    suppressMessages(get_forecast_counts(ex, by = "model"))
  )
})
```

We have a similar issue in when running the following piece of code from the scoringutils.Rmd Vignette:
```
suppressMessages(score(example_point)) %>%
  summarise_scores(by = "model", na.rm = TRUE)
```
The issue here is that the output of `score()` is still of class `forecast_sample` (which it shouldn't be). I suggest addressing this in a different PR though (already started working on it). 

The current suggestion defines an extra function, `print_forecast_info()`. I think in principle we don't need the function. We could also just do 
```
print.forecast_binary <- function(x, ...) {yourcode}
```
and 
```
print.forecast_quantile <- print.forecast_binary
``` 
etc. Probably even better, we should consider introducing a super-class, `forecast` or `scoringutils` that all of these objects have. This could then be used for printing and we would only have to define a single printing function. Maybe it's worth addressing this in a separate issue/PR and going with the first option for now. What do you think, @seabbs?

I would personally prefer to keep the protected columns out of the printing method for now and introduce them later. I think it is generally nice to inform about protected columns. However, I would prefer to add them at a later point in time because we don't really have written guidance on what a protected column actually is. At the moment it is "something that functions aren't allowed to touch and where you need to be careful to rename them". But there are a lot of uncertainties. What are our protected columns exactly? Is a score a protected column? I think we need something written down that explains what a protected column actually is before we can inform users about the protected columns. 

Summarising this wall of text, my suggestions are
- updating line 63 in `available_forecasts.R` + adding a test to make sure that the problem doesn't occur in the future
- removing `print_forecast_info()`
- fixing the class that the output of `score()` has - I can take care of that
- removing the protected columns from the print method for now and reintroducing them later once we have a clear agreed understanding of what a protected column is

What do others think? I think the overall PR is really good, as always lots of tiny details to deal with. Thanks a lot again.

seabbs commented:
> I suggest addressing this in a different PR though (already started working on it).

i.e. no action needed here for you @toshiakiasakura 

>  This could then be used for printing and we would only have to define a single printing function. Maybe it's worth addressing this in a separate issue/PR and going with the first option for now. What do you think

Yeah I think getting something simple off the ground makes sense and then thinking about merging in a future issue/PR would be good. I can see a good argument for why we would want a meta-class or classes (for example for categorical forecasts (which have nominal, ordinal, and binary in them) etc.) but again yes a problem for another day IMO.


So for @toshiakiasakura / this PR we have: 

- [ ] updating line 63 in available_forecasts.R + adding a test to make sure that the problem doesn't occur in the future
- [ ] removing print_forecast_info() and code directly into the methods
- [ ] removing the protected columns from the print method for now and reintroducing them later once we have a clear agreed understanding of what a protected column i

seabbs commented on R/utils.R:
```diff
@@ -200,3 +200,61 @@ ensure_data.table <- function(data) {
+  print(forecast_units)
+
+  cat("\n")
+  print(as.data.table(dat), ...)
```
We should be able to use next method here to make tis a bit more generalisable (see nextmethod in here: http://adv-r.had.co.nz/S3.html)

toshiakiasakura replied:
Revised in [e368ebf](https://github.com/epiforecasts/scoringutils/pull/592/commits/e368ebfa92614fe9155eeef6da3e28a18be0b3c6)

nikosbosse commented:
@toshiakiasakura How are you feeling about this at the moment? Happy to help get this PR over the finish line!

toshiakiasakura commented:
Thanks for a lot of suggestions. I will try to work on following the checklist. I agree with most. 

One reason I separately prepare `print_forecast_info' is that in my mind, in the later phase, I am thinking of introducing the validating results in the print result. For that purpose, we need a bit different implementation for each print object, and then allow the code to be extended differently. 

For the simplicity, should I still at this stage amend the code to remove `print_forecast_info`?

seabbs commented:
> For the simplicity, should I still at this stage amend the code to remove print_forecast_info?

Yes I think that makes sense for this PR

toshiakiasakura commented:
I've almost done with the updates. 

> * [x]  updating line 63 in available_forecasts.R + adding a test to make sure that the problem doesn't occur in the future
I added

For testing `get_forecast_counts`, I simply added one line to check the returning object is exactly matched with the `data.table` class with `expect_s3_class`. 

> * [x]  removing print_forecast_info() and code directly into the methods
> * [x]  removing the protected columns from the print method for now and reintroducing them later once we have a clear agreed understanding of what a protected column
> * [x]  Use nextMethod vs using print(as.data.table)
> * [ ]  Clean up linting issue

> * [x]  Ideally add some kind of minimal testing of the new functionality

Added the simple test to check the print function outputs the "Forecast type" and "Forecast units". 

Currently output is like this. 
```R
example_quantile %>%
  set_forecast_unit(c("location", "target_end_date", "target_type", "horizon", "model")) %>%
  as_forecast() %>%
  add_coverage() %>%
  print
#> Forecast type:
#> [1] "quantile"
#> 
#> Score columns:
#> [1] "interval_coverage"           "interval_coverage_deviation"
#> [3] "quantile_coverage"           "quantile_coverage_deviation"
#> 
#> Forecast units:
#> [1] "location"        "target_end_date" "target_type"     "horizon"        
#> [5] "model"          
#> 
#>        observed quantile predicted location target_end_date target_type horizon
#>     1:   127300       NA        NA       DE      2021-01-02       Cases      NA


```

nikosbosse commented on R/utils.R:
```diff
@@ -200,3 +200,47 @@ ensure_data.table <- function(data) {
+#' including "Forecast type", "Score columns",
+#' "Forecast unit".
+#' @export
+#' @keywords print-forecast
```
Ah the pkgdown test is complaining about this line because the topic must be included in the reference list in _pkgdown.yaml. I suggest changing it to one of the keywords we already have and revisiting later when we overhaul the pkgdown index. 

```suggestion
#' @keywords check-forecast
```

nikosbosse commented on R/utils.R:
```diff
@@ -200,3 +200,47 @@ ensure_data.table <- function(data) {
+#' @export
+#' @keywords print-forecast
+#' @examples
+#' print(example_quantile %>% as_forecast)
```
This line produces an error since magrittr is not automatically part of the namespace. 

I would suggest to simply replace this by 
```suggestion
#' as_forecast(example_quantile)
```

toshiakiasakura replied:
Since this is a print function, we need to write as `print(as_forecast(example_quantile))` ?

toshiakiasakura replied:
Done in the commit [d3a75cd](https://github.com/epiforecasts/scoringutils/pull/592/commits/d3a75cda48646d206673faba8ddc90e8510cf536)

nikosbosse commented on R/utils.R:
```diff
@@ -200,3 +200,47 @@ ensure_data.table <- function(data) {
   return(data)
 }
+
+#' @title Print information about a `forecast` object.
```
Suggest making this title case for consistency with (most...) other functions

```suggestion
#' @title Print Information About A Forecast Object
```

nikosbosse commented on R/utils.R:
```diff
@@ -200,3 +200,47 @@ ensure_data.table <- function(data) {
+    print(score_cols)
+  }
+
+  cat("\nForecast units:\n")
```
The manuscript uses the singular, so I suggest we do it here as well: 
```suggestion
  cat("\nForecast unit:\n")
```

nikosbosse commented on R/utils.R:
```diff
@@ -200,3 +200,47 @@ ensure_data.table <- function(data) {
+#' @keywords print-forecast
+#' @examples
+#' print(example_quantile %>% as_forecast)
+print.forecast_binary <- function(dat, ...) {
```
Methods must have the same arguments as the generic (otherwise R CMD check produces a warning, see https://adv-r.hadley.nz/s3.html#s3-arguments). I therefore suggest changing the function arguments to 

```suggestion
print.forecast_binary <- function(x, ...) {
```

even though x is not really the clearest argument 🤷. But it's enforced by R CMD check.

nikosbosse replied:
Also I think `...` needs documentation. Suggest something like
```
#' param ... additional arguments for [print()]
```

toshiakiasakura replied:
I added several arguments in addition to `...` following `print.scoringutils_check` and done in [d3a75cd](https://github.com/epiforecasts/scoringutils/pull/592/commits/d3a75cda48646d206673faba8ddc90e8510cf536).

nikosbosse commented on R/utils.R:
```diff
@@ -200,3 +200,54 @@ ensure_data.table <- function(data) {
+#' @param x An object of class 'forecast_*' object as produced by
+#' `as_forecast()`
+#' @param ... additional arguments for [print()]
+#' 
```
```suggestion
```

nikosbosse replied:
There is a setting in Rstudio where you can automatically remove trailing whitespaces on every save. I'm using that and it's really nice.

toshiakiasakura replied:
I am using VS code, and now I turn on the function of automatically removing all trailing whitespace. Thanks for the suggestion.

nikosbosse commented:
Thanks a lot, @toshiakiasakura! 
I made a small suggested change that just removes a line of code with a whitespace. The linter is complaining about superfluous whitespaces. 

I'm not sure why the snapshot tests are failing. I checked out the code locally and tests passed there. Did they pass locally on your end? I'll investigate. 

As an aside, we might also want to think about 
- adding a test for the print method, e.g. using `expect_output`
- failing gracefully. In the current version, the user will get an error if something like `get_forecast_unit()` fails for whatever reason. Ideally, the function should at least print the data.table, regardless of any other errors so the user can inspect it.

toshiakiasakura commented:
Sure, I will check later the linting problem.  

For testing, I implemented a very simple test in https://github.com/epiforecasts/scoringutils/commit/db0323fd33bc3beac05ebd1fddcb241f540a8a0c . 
Do you think it is insufficient?

toshiakiasakura commented:
> I'm not sure why the snapshot tests are failing. I checked out the code locally and tests passed there. Did they pass locally on your end? I'll investigate.

Yes, the test is unfortunately passed even though there is one trailing white space... 

> adding a test for the print method, e.g. using expect_output

I added. 

> failing gracefully. In the current version, the user will get an error if something like get_forecast_unit() fails for whatever reason. Ideally, the function should at least print the data.table, regardless of any other errors so the user can inspect it.

@nikosbosse  For this part, there is also a possibility that `forecast_object` is collapsed and `get_forecast_unit` does not work. 
My simple suggestion is to rely on `validate_forecast` to check the object, and if an error is present, print that error. If that test is passed, print the information of the `forecast` object as currently implemented. To achieve this, I am planning to write in the following way. What do you think about it?

```R
res <- try(validate_forecast(x[, c("observed", "predicted")] ), silent=TRUE)
if (class(res) == "try-error"){
    message(res)
} else {
    # current print contents.
}
```
or cause the error (and stop the flow) by putting `validate_forecast(x)` before the current print contents?

toshiakiasakura commented:
For failing gracefully, I am wondering what case you expect the breakage of the `forecast` object. As far as I understand, broken usage is limited if the `forecast` object is kept. I guess converting from a `forecast` object to a `data.table` sometimes happens but that does not matter for this print functionality. So I prefer to keep the code simple.

nikosbosse commented:
@toshiakiasakura What I mean is that, for example, the following produces an error, because the object can't get printed anymore because `get_forecast_type()`, which is called by `print.forecast_()` requires a column "observed". 

```
library(dplyr)
ex <- example_quantile |>
  as_forecast()

ex |> 
  select(-observed) 
```
Ideally, there should be something like an error (or a warning), but the data.table would still print. 

@seabbs any thoughts on whether this should be part of this PR or a separate PR? I lean towards addressing it in this PR as it seems quite important to me that users can still print their output even if something is wrong. But I appreciate that this would mean adding even more to this PR, for which I apologise, @toshiakiasakura!

seabbs commented:
> @seabbs any thoughts on whether this should be part of this PR or a separate PR? I lean towards addressing it in this PR as it seems quite important to me that users can still print their output even if something is wrong.

I am leaning towards addressing this in its own issue/PR as the suggestion above is already functional and we have had a fair few rounds of review already and I imagine everyone would like to see this on `main` soon!

nikosbosse commented on NEWS.md:
```diff
@@ -45,6 +45,7 @@ The update introduces breaking changes. If you want to keep using the older vers
 - Added unit tests for `interval_coverage_quantile()` and `interval_coverage_dev_quantile()` in order to make sure that the functions provide the correct warnings when insufficient quantiles are provided.
 - Documentation pkgdown pages are now created both for the stable and dev versions.
 - Output columns for pairwise comparisons have been renamed to contain the name of the metric used for comparing.
+- Added `print()` on a `forecast` object to print out diagnostic messages. 
```
```suggestion
- Added a method for `print()` that prints out additional information for `forecast` objects. 
```

seabbs reviewed (approved):
LGTM. Thanks for this @toshiakiasakura !! I suggest we also open another issue to think about if we could make this even prettier using `{cli}`.

nikosbosse commented:
Perfect, merging now. Thanks a lot @toshiakiasakura! I think this is a really visible and important improvement to users.
